src/pu-main.c

```diff
@@ -5,168 +5,258 @@
-                    "%s", "Excess values in remaining arguments!");
+    if (g_strcmp0(device_path, "") <= 0) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "No device specified");
         return FALSE;
     }
 
+    if (!pu_is_drive(device_path)) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Device '%s' is not a drive", device_path);
+        return FALSE;
+    }
+
+    if (!pu_device_mounted(device_path, &is_mounted, error)) {
+        g_prefix_error(error, "Failed checking if device is in use: ");
+        return FALSE;
+    }
+
+    if (is_mounted) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Device '%s' is in use", device_path);
+        return FALSE;
+    }
+
+    if (!pu_package_mount(package_path, &mount_path, error))
+        return FALSE;
+
+    config_path = pu_package_get_layout_file(mount_path, error);
+    config = pu_config_new_from_file(config_path, error);
+    if (config == NULL) {
+        g_prefix_error(error, "Failed creating configuration object for file '%s': ",
+                       config_path);
+        return error_out(mount_path);
+    }
+    api_version = pu_config_get_api_version(config);
+    if (api_version > PARTUP_VERSION_MAJOR) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "API version %d of configuration file is not compatible "
+                    "with program version %d", api_version, PARTUP_VERSION_MAJOR);
+        return error_out(mount_path);
+    }
+
+    emmc = pu_emmc_new(device_path, config, mount_path,
+                       arg_install_skip_checksums, error);
+    if (emmc == NULL) {
+        g_prefix_error(error, "Failed parsing eMMC info from config: ");
+        return error_out(mount_path);
+    }
+    if (!pu_flash_init_device(PU_FLASH(emmc), error)) {
+        g_prefix_error(error, "Failed initializing device: ");
+        return error_out(mount_path);
+    }
+    if (!pu_flash_setup_layout(PU_FLASH(emmc), error)) {
+        g_prefix_error(error, "Failed setting up layout on device: ");
+        return error_out(mount_path);
+    }
+    if (!pu_flash_write_data(PU_FLASH(emmc), error)) {
+        g_prefix_error(error, "Failed writing data to device: ");
+        pu_umount_all(device_path, NULL);
+        return error_out(mount_path);
+    }
+
+    return pu_package_umount(mount_path, error);
+}
+
+static gboolean
+cmd_package(PuCommandContext *context,
+            GError **error)
+{
+    g_autofree gchar *input_files = NULL;
+    g_autofree gchar *package = NULL;
+    gchar **args;
+
+    args = pu_command_context_get_args(context);
+
+    if (!g_path_is_absolute(args[0]) && arg_package_directory)
+        package = g_build_filename(g_get_current_dir(), args[0], NULL);
+    else
+        package = g_strdup(args[0]);
+
+    for (gsize i = 1; args[i] != NULL; i++)
+        input_files = g_strconcat(input_files, " ", args[i], NULL);
```

Comment:
`input_files` must not be `NULL` but is for the first loop iteration.

Reply:
Good find! I'll remove the for-loop using `g_strconcat()` in favor of simply retrieving the input files from the offset of the `args` array with `&args[1]`.